satellite/satellitedb/overlaycache.go

```diff
@@ -27,100 +23,53 @@ type overlaycache struct {
+		  AND audit_count >= ?
+		  AND audit_success_ratio >= ?
+		  AND uptime_count >= ?
+		  AND audit_uptime_ratio >= ?
+		`, int(criteria.Type), criteria.FreeBandwidth, criteria.FreeDisk,
+		criteria.AuditCount, criteria.AuditSuccessRatio, criteria.UptimeCount, criteria.UptimeSuccessRatio,
+	)
 }
 
-// FilterNodes looks up nodes based on reputation requirements
-func (cache *overlaycache) FilterNodes(ctx context.Context, req *overlay.FilterNodesRequest) ([]*pb.Node, error) {
-	reputableNodeAmount := req.MinNodes
-	if reputableNodeAmount <= 0 {
-		reputableNodeAmount = req.Opts.GetAmount()
-	}
-
-	getReputableReq := &getNodesRequest{
-		minReputation:         req.MinReputation,
-		freeBandwidth:         req.Opts.GetRestrictions().FreeBandwidth,
-		freeDisk:              req.Opts.GetRestrictions().FreeDisk,
-		excluded:              req.Opts.ExcludedNodes,
-		reputableNodeAmount:   reputableNodeAmount,
-		newNodeAuditThreshold: req.NewNodeAuditThreshold,
-	}
-
-	reputableNodes, err := cache.getReputableNodes(ctx, getReputableReq)
-	if err != nil {
-		return nil, err
-	}
-
-	newNodeAmount := int64(float64(reputableNodeAmount) * req.NewNodePercentage)
-
-	getNewReq := &getNodesRequest{
-		freeBandwidth:         req.Opts.GetRestrictions().FreeBandwidth,
-		freeDisk:              req.Opts.GetRestrictions().FreeDisk,
-		excluded:              req.Opts.ExcludedNodes,
-		newNodeAmount:         newNodeAmount,
-		newNodeAuditThreshold: req.NewNodeAuditThreshold,
-	}
-
-	newNodes, err := cache.getNewNodes(ctx, getNewReq)
-	if err != nil {
-		return nil, err
-	}
-
-	var allNodes []*pb.Node
-	allNodes = append(allNodes, reputableNodes...)
-	allNodes = append(allNodes, newNodes...)
-
-	if int64(len(reputableNodes)) < reputableNodeAmount {
-		err := status.Errorf(codes.ResourceExhausted, fmt.Sprintf("requested %d reputable nodes, only %d reputable nodes matched the criteria requested",
-			reputableNodeAmount, len(reputableNodes)))
-		return allNodes, err
-	}
-
-	return allNodes, nil
+func (cache *overlaycache) SelectNewNodes(ctx context.Context, count int, criteria *overlay.NewNodeCriteria) ([]*pb.Node, error) {
+	return cache.queryFilteredNodes(ctx, criteria.Excluded, count, `
+		WHERE node_type == ? AND free_bandwidth >= ? AND free_disk >= ?
+		  AND audit_count < ?
+	`, int(criteria.Type), criteria.FreeBandwidth, criteria.FreeDisk,
+		criteria.AuditThreshold,
+	)
 }
 
-func (cache *overlaycache) getReputableNodes(ctx context.Context, req *getNodesRequest) ([]*pb.Node, error) {
-	rows, err := cache.findReputableNodesQuery(ctx, req)
-	if err != nil {
-		return nil, err
+func (cache *overlaycache) queryFilteredNodes(ctx context.Context, excluded []storj.NodeID, count int, safeQuery string, args ...interface{}) (_ []*pb.Node, err error) {
+	if count == 0 {
+		return nil, nil
 	}
-	defer func() {
-		err = utils.CombineErrors(err, rows.Close())
-	}()
 
-	reputableNodes, err := sqlRowsToNodes(rows)
-	if err != nil {
-		return nil, err
+	safeExcludeNodes := ""
```

Comment:
nit - maybe should be `safeExcludedNodes` or needs a comment to explain why it's safe